flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/ZkLockFactory.java

```diff
@@ -0,0 +1,197 @@
+public class ZkLockFactory implements TriggerLockFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkLockFactory.class);
+
+  private static final String LOCK_BASE_PATH = "/iceberg/flink/maintenance/locks";
+
+  private final String connectString;
+  private final String lockId;
+  private final int sessionTimeoutMs;
+  private final int connectionTimeoutMs;
+  private final ExponentialBackoffRetryWrapper retryWrapper;
+  private transient CuratorFramework client;
+
+  /**
+   * Create Zookeeper lock factory
+   *
+   * @param connectString Zookeeper connection string
+   * @param lockId Lock ID to distinguish different jobs
+   * @param sessionTimeoutMs Session timeout in milliseconds
+   * @param connectionTimeoutMs Connection timeout in milliseconds
+   */
+  public ZkLockFactory(
+      String connectString,
+      String lockId,
+      int sessionTimeoutMs,
+      int connectionTimeoutMs,
+      ExponentialBackoffRetryWrapper retryWrapper) {
+    Preconditions.checkNotNull(connectString, "Zookeeper connection string cannot be null");
+    Preconditions.checkNotNull(lockId, "Lock ID cannot be null");
+    this.connectString = connectString;
+    this.lockId = lockId;
+    this.sessionTimeoutMs = sessionTimeoutMs;
+    this.connectionTimeoutMs = connectionTimeoutMs;
+    this.retryWrapper = retryWrapper;
+  }
+
+  @Override
+  public void open() {
+    this.client =
+        CuratorFrameworkFactory.builder()
+            .connectString(connectString)
+            .sessionTimeoutMs(sessionTimeoutMs)
+            .connectionTimeoutMs(connectionTimeoutMs)
+            .retryPolicy(retryWrapper.getRetryPolicy())
+            .build();
+    client.start();
+
+    try {
+      if (!client.blockUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)) {
+        throw new IllegalStateException("Connection to Zookeeper timed out");
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted while connecting to Zookeeper", e);
+    }
+  }
+
+  @Override
+  public Lock createLock() {
+    return new ZkLock(client, LOCK_BASE_PATH + "/maintenance/" + lockId);
+  }
+
+  @Override
+  public Lock createRecoveryLock() {
+    return new ZkLock(client, LOCK_BASE_PATH + "/recovery/" + lockId);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (client != null) {
+      client.close();
+    }
+  }
+
+  /** Zookeeper lock implementation */
+  private static class ZkLock implements Lock {
+    private final String lockPath;
+    private final CuratorFramework client;
+
+    private ZkLock(CuratorFramework client, String lockPath) {
+      this.client = client;
+      this.lockPath = lockPath;
+    }
+
+    @Override
+    public boolean tryLock() {
+      if (isHeld()) {
+        LOG.info("Lock is already held for {}", this);
+        return false;
+      }
+
+      String newInstanceId = UUID.randomUUID().toString();
```

Comment:
Do we need this `newInstanceId` at all?
With the `JDBCLockFactory` we needed it because we had to handle retries.
ZK retries are handled by the curator framework, so I think we can just rely on that, and create/delete nodes.
Do you have different experience with ZK?